added rbac manifests generation pkg and CLI

droot opened the pull request:
Note: need feedback specifically about the pkg and command's location. 

Added `rbac` new commandline tool and `rbac` pkg under `pkg/generate/rbac` for generating RBAC manifests.  It scans the Go source files under given input-dir and parses the RBAC annotations and generates the RBAC manifests. The annotations could be anywhere in the Go files not necessarily tied to a Type or Function (Thats one reason `gengo` wasn't useful). Parser uses `Go` standard packages (ast, parser) instead of `gengo` because requirements were very simple. 

Realized that we will have to generate RBAC annotations while scaffolding the resources so that this tool can generate RBAC manifests.

mengqiy commented on pkg/generate/rbac/manifests.go:
```diff
@@ -0,0 +1,125 @@
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-rolebinding", o.Name),
+			Namespace: fmt.Sprintf("%s-system", o.Name),
```
IIRC namespace will be ignored.
I think you can drop this line.

droot replied:
Done.

mengqiy commented on pkg/generate/rbac/parser.go:
```diff
@@ -0,0 +1,145 @@
+		case "groups":
+			normalized := []string{}
+			for _, v := range values {
+				if v == "core" {
```
```
// +rbac:groups="",resources=pods,verbs=get;list;watch;delete
```
IMO using an empty string for core group should also be valid.
WDYT?

droot replied:
Yes, it is. Current code is supporting both "" and "core".
Not dropping the support for "core" yet because old projects might have them, so it might help with the migration.

mengqiy commented on pkg/generate/rbac/parser.go:
```diff
@@ -0,0 +1,145 @@
+				}
+			}
+			if strings.HasPrefix(comment, "+kubebuilder:rbac") {
+				if ann := getAnnotation(comment, "kubebuilder:rbac", ":"); ann != "" {
```
Are we going support 2 annotations forever?
If not, we should print warning for the one that we eventually will drop support.

droot replied:
We might be able to drop the "rbac" annotation completely for new projects. Though keeping it around will help with the migration. 
Would like to deprecate it in separate PR (along with other deprecations).

mengqiy commented on pkg/generate/rbac/parser_test.go:
```diff
@@ -0,0 +1,82 @@
+	)
+
+	// RBAC annotation with kubebuilder prefix
+	// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;delete
```
We may want to have more test cases to cover
1) core group
2) using `""`

droot replied:
Added.

lichuqiang commented on cmd/rbac/main.go:
```diff
@@ -0,0 +1,58 @@
+		Short: "Generates RBAC manifests",
+		Long: `Generate RBAC manifests from the RBAC annotations in Go source files.
+Usage:
+# rbac generate [--name manager] [--input-dir input_dir] [--output-dir output_dir]
```
I think the sub command "generate" here did not take effect. i.e. ”rbac [--name manager] [--input-dir input_dir] [--output-dir output_dir]“ can also work.

Maybe add it as a sub command explicitly

droot replied:
Good pt. Created a new command `controller-gen` and added `rbac` as a subcommand. So here is the plan:

we will invoke `controller-gen all` using `go generate` directive from the generated project. `all` subcommand will trigger all the generators for a project.

Liujingfang1 commented on pkg/generate/rbac/manifests.go:
```diff
@@ -0,0 +1,124 @@
+	InputDir  string
+	OutputDir string
+	Name      string
+	Labels    map[string]string
```
Where is Labels read in? I don't see a flag for it

droot replied:
Didn't implement the flag part in this version :). Will address it a separate PR.

droot commented:
Addressed the command. Main change is created a new command controller-gen and added rbac as a subcommand. So here is the plan:

we will invoke controller-gen all using go generate directive from the generated project. all subcommand will trigger all the generators for a project.

@mengqiy PTAL.
